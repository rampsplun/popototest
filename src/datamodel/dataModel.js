var dataModel = {};

dataModel.idGen = 0;

dataModel.generateId = function () {
    return dataModel.idGen++;
};

dataModel.nodes = [];
dataModel.links = [];

dataModel.getRootNode = function () {
    return dataModel.nodes[0];
};

export default dataModel