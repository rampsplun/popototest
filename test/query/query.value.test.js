import provider from "../../src/provider/provider";
import query from '../../src/query/query.js'
import dataModel from "../../src/datamodel/dataModel";

describe("root only", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            }
        ];

        dataModel.links = [];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("generateNodeValueQuery", () => {
        var generateNodeCountQuery = query.generateNodeValueQuery(dataModel.nodes[0]);
        expect(generateNodeCountQuery.statement).toMatchSnapshot();
        expect(generateNodeCountQuery.parameters).toMatchSnapshot();
    });
});

describe("COLLECT_RELATIONS_WITH_VALUES", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Label1",
                internalLabel: "label1",
                type: 1,
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Label1": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };

        query.COLLECT_RELATIONS_WITH_VALUES = true;
    });

    afterEach(() => {
        delete provider.node.Provider;
        query.COLLECT_RELATIONS_WITH_VALUES = false;
    });

    test("generateNodeValueQuery", () => {
        var generateNodeCountQuery = query.generateNodeValueQuery(dataModel.nodes[1]);
        expect(generateNodeCountQuery.statement).toMatchSnapshot();
        expect(generateNodeCountQuery.parameters).toMatchSnapshot();
    });
});

describe("one very long-branch with value leaf and reverse order", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Label1",
                internalLabel: "label1",
                type: 1,
            },
            {
                id: 2,
                label: "Label2",
                isParentRelReverse: true,
                internalLabel: "label2",
                type: 1,
            },
            {
                id: 3,
                label: "Label3",
                internalLabel: "label3",
                value: [{label: "Label3", attributes: {Nid: "Vid3", Nname: "Vname3"}}],
                isNegative: true,
                type: 1,
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO3",
                source: dataModel.nodes[2],
                target: dataModel.nodes[3],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Label1": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            },
            "Label2": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            },
            "Label3": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("generateNodeValueQuery", () => {
        var generateNodeValueQuery = query.generateNodeValueQuery(dataModel.nodes[2]);
        expect(generateNodeValueQuery.statement).toMatchSnapshot();
        expect(generateNodeValueQuery.parameters).toMatchSnapshot();

        generateNodeValueQuery = query.generateNodeValueQuery(dataModel.nodes[3]);
        expect(generateNodeValueQuery.statement).toMatchSnapshot();
        expect(generateNodeValueQuery.parameters).toMatchSnapshot();
    });
});

describe("Predefined constraints Neo4j id generation", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node1",
                type: 1
            },
            {
                id: 2,
                label: "Node",
                internalLabel: "node2",
                value: [{label: "Node", internalID: 102, attributes: {Nname: "name2"}}],
                type: 1,
            },
            {
                id: 3,
                label: "Node",
                internalLabel: "node3",
                type: 1,
                value: [
                    {label: "Node", internalID: 1031, attributes: {Nname: "name31"}},
                    {label: "Node", internalID: 1032, attributes: {Nname: "name32"}}
                ],
            }
        ];

        dataModel.links = [
            {
                id: 0,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 1,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            },
            {
                id: 0,
                label: "LINKED_TO3",
                source: dataModel.nodes[0],
                target: dataModel.nodes[3],
                type: 1
            },
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: [query.NEO4J_INTERNAL_ID],
                constraintAttribute: query.NEO4J_INTERNAL_ID,
                getPredefinedConstraints: function () {
                    return ["$identifier.x > 18"];
                }
            },
            "Node": {
                returnAttributes: ["Nname", query.NEO4J_INTERNAL_ID],
                constraintAttribute: query.NEO4J_INTERNAL_ID,
                getPredefinedConstraints: function () {
                    return ["$identifier.y < 20"];
                }
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("predefined constraints", () => {
        var generateNodeValueQuery = query.generateNodeValueQuery(dataModel.nodes[0]);
        expect(generateNodeValueQuery.statement).toMatchSnapshot();
        expect(generateNodeValueQuery.parameters).toMatchSnapshot();
    });

    test("Neo4j id", () => {
        var generateNodeValueQuery = query.generateNodeValueQuery(dataModel.nodes[2]);
        expect(generateNodeValueQuery.statement).toMatchSnapshot();
        expect(generateNodeValueQuery.parameters).toMatchSnapshot();
    });

    test("Neo4j id", () => {
        var generateNodeValueQuery = query.generateNodeValueQuery(dataModel.nodes[3]);
        expect(generateNodeValueQuery.statement).toMatchSnapshot();
        expect(generateNodeValueQuery.parameters).toMatchSnapshot();
    });
});