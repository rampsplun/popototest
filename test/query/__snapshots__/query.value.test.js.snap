// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COLLECT_RELATIONS_WITH_VALUES generateNodeValueQuery 1`] = `"MATCH (root:\`Root\`), (root:\`Root\`)-[r0:\`LINKED_TO1\`]->(label1:\`Label1\`) RETURN label1.Nname AS Nname, label1.Nid AS Nid, count(DISTINCT root.id) AS count, COLLECT(r0) AS incomingRels ORDER BY count DESC LIMIT 100"`;

exports[`COLLECT_RELATIONS_WITH_VALUES generateNodeValueQuery 2`] = `Object {}`;

exports[`Predefined constraints Neo4j id generation Neo4j id 1`] = `"MATCH (root:\`Root\`), (node1:\`Node\`)-[r0:\`LINKED_TO2\`]->(node2:\`Node\`), (root:\`Root\`)-[r1:\`LINKED_TO3\`]->(node3:\`Node\`), (root:\`Root\`)-[r2:\`LINKED_TO1\`]->(node1:\`Node\`) WHERE root.x > 18 AND node2.y < 20 AND node3.y < 20 AND ID(node3) IN $node3_internalID AND node1.y < 20 RETURN node2.Nname AS Nname, ID(node2) AS NEO4JID, count(DISTINCT ID(root)) AS count ORDER BY count DESC LIMIT 100"`;

exports[`Predefined constraints Neo4j id generation Neo4j id 2`] = `
Object {
  "node3_internalID": Array [
    1031,
    1032,
  ],
}
`;

exports[`Predefined constraints Neo4j id generation Neo4j id 3`] = `"MATCH (root:\`Root\`), (node1:\`Node\`)-[r0:\`LINKED_TO2\`]->(node2:\`Node\`), (root:\`Root\`)-[r1:\`LINKED_TO3\`]->(node3:\`Node\`), (root:\`Root\`)-[r2:\`LINKED_TO1\`]->(node1:\`Node\`) WHERE root.x > 18 AND node2.y < 20 AND ID(node2) = $node2_internalID AND node3.y < 20 AND node1.y < 20 RETURN node3.Nname AS Nname, ID(node3) AS NEO4JID, count(DISTINCT ID(root)) AS count ORDER BY count DESC LIMIT 100"`;

exports[`Predefined constraints Neo4j id generation Neo4j id 4`] = `
Object {
  "node2_internalID": 102,
}
`;

exports[`Predefined constraints Neo4j id generation predefined constraints 1`] = `"MATCH (root:\`Root\`), (node1:\`Node\`)-[r0:\`LINKED_TO2\`]->(node2:\`Node\`), (root:\`Root\`)-[r1:\`LINKED_TO3\`]->(node3:\`Node\`), (root:\`Root\`)-[r2:\`LINKED_TO1\`]->(node1:\`Node\`) WHERE root.x > 18 AND node2.y < 20 AND ID(node2) = $node2_internalID AND node3.y < 20 AND ID(node3) IN $node3_internalID AND node1.y < 20 RETURN ID(root) AS NEO4JID, count(DISTINCT ID(root)) AS count ORDER BY count DESC LIMIT 100"`;

exports[`Predefined constraints Neo4j id generation predefined constraints 2`] = `
Object {
  "node2_internalID": 102,
  "node3_internalID": Array [
    1031,
    1032,
  ],
}
`;

exports[`one very long-branch with value leaf and reverse order generateNodeValueQuery 1`] = `"MATCH (root:\`Root\`), (label1:\`Label1\`)<-[r0:\`LINKED_TO2\`]-(label2:\`Label2\`), (root:\`Root\`)-[r1:\`LINKED_TO1\`]->(label1:\`Label1\`) WHERE (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Label1\`)<-[:\`LINKED_TO2\`]-(:\`Label2\`)-[:\`LINKED_TO3\`]->(:\`Label3\`{Nid:$label3_Nid}))) RETURN label2.Nname AS Nname, label2.Nid AS Nid, count(DISTINCT root.id) AS count ORDER BY count DESC LIMIT 100"`;

exports[`one very long-branch with value leaf and reverse order generateNodeValueQuery 2`] = `
Object {
  "label3_Nid": "Vid3",
}
`;

exports[`one very long-branch with value leaf and reverse order generateNodeValueQuery 3`] = `"MATCH (root:\`Root\`), (label2:\`Label2\`)-[r0:\`LINKED_TO3\`]->(label3:\`Label3\`), (label1:\`Label1\`)<-[r1:\`LINKED_TO2\`]-(label2:\`Label2\`), (root:\`Root\`)-[r2:\`LINKED_TO1\`]->(label1:\`Label1\`) WHERE (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Label1\`)<-[:\`LINKED_TO2\`]-(:\`Label2\`)-[:\`LINKED_TO3\`]->(:\`Label3\`{Nid:$label3_Nid}))) RETURN label3.Nname AS Nname, label3.Nid AS Nid, count(DISTINCT root.id) AS count ORDER BY count DESC LIMIT 100"`;

exports[`one very long-branch with value leaf and reverse order generateNodeValueQuery 4`] = `
Object {
  "label3_Nid": "Vid3",
}
`;

exports[`root only generateNodeValueQuery 1`] = `"MATCH (root:\`Root\`) RETURN root.name AS name, root.id AS id, count(DISTINCT root.id) AS count ORDER BY count DESC LIMIT 100"`;

exports[`root only generateNodeValueQuery 2`] = `Object {}`;
